Add element columns to sort table

lukasturcani opened the pull request:
# Note

I have a question:

## Should the element be automatically wrapped in an `Element.el?`

For example, this code

```elm
            , Widget.elementColumn
                { title = ""
                , value = \task ->
                        Widget.switch
                            (Material.switch Material.darkPalette)
                            { description = "active"
                            , onPress = Just <| ToggleActive task.index
                            , active = task.active
                            }
                        
                , width = Element.fill
                }

```

leads to

![image](https://user-images.githubusercontent.com/17861210/146683858-f9616ed1-250b-4964-b006-782d3d2ae9f7.png)


I have to do 

```elm

            , Widget.elementColumn
                { title = ""
                , value = \task ->
                    Element.el
                        []
                        (Widget.switch
                            (Material.switch Material.darkPalette)
                            { description = "active"
                            , onPress = Just <| ToggleActive task.index
                            , active = task.active
                            }
                        )
                , width = Element.fill
                }


```


to make it look right

![image](https://user-images.githubusercontent.com/17861210/146683842-7d1b837b-2bae-4204-b50e-f80404632e05.png)


I could edit the PR so that `Element.el` is used automatically, and the user does not have to write it,  but I'm wondering if this can cause issues in other cases?

# Description

* I extended the `sortTable` to allow the embedding of arbitrary `Element` into columns.
* Made`Column` and `ColumnType` types polymorphic on a `msg` type variable. This allows them to hold data of type `Element msg`
* Updated type definitions to use `Column a msg` and `ColumnType a msg`
* Added a new constructor for `ColumnType`, `ElementType`, which is used to construct columns which hold `Element msg` data
* Added a constructor function `elementColumn` for `Column` instances, which have values of type `Element msg`
* Updated the definition of `sortTable` to handle `ElementColumn` data. `ElementColumn` data is not sortable.
* Refactored the `view` in `sortTable`.  See https://github.com/Orasund/elm-ui-widgets/pull/85/files#r771978681
* Updated doc of `sortTable` to include example of `elementColumn` use

**Pull request solving issue #84.**

_Please only open Pull Requests for existing issues._

# Checklist for new Widget

* [x] Checked out the Material design specification of the widget
* [x] Added constructors, Widgets Type and Style Type in `src/Internal/[Your Widget].elm`
* [ ] Added styles in `src/Internal/Material/[Your Widget].elm`
* [ ] In `src/Widget.Material.elm`:
  * [ ] Linked each style to its representative in `src/Internal/Material/[Your Widget].elm`
  * [ ] Added Documentation
* [x] In `src/Widget.elm`:
  * [x] Added a copy of the Widget Type
  * [ ] Added a copy of the Style Type
  * [x] Linked each constructor to its representative in `src/Internal/[You Widget].elm`
  * [x] Replaced the Widget Type in the type signiture of each constructor with its definition
  * [x] Added Documentation with a small example
  * [x] run `elm-verify-examples && elm-test` to test the example

**Optional:**
* [x] Added a Page in `explorer/src/Page/[Your Widget].elm` and added it to `explorer/src/Main.elm`

lukasturcani commented on src/Internal/SortTable.elm:
```diff
@@ -180,20 +195,32 @@ sortTable style model =
                                 IntColumn { value, toString } ->
-                                    value >> toString
+                                    value
+                                        >> toString
```
I moved this shared code into the code for each case, because `ElementColumn` case cannot share the code, since it's already an `Element`.

Orasund replied:
Please rename `ElementColumn` into `CustomColumn`.

Orasund commented:
Currently, this would be a major change. I'd like it to instead be a minor one, so please instead of changing the type signature of a function, create a V2 version instead. So `SortTableV2, ColumnV2, sortTableV2, floatColumnV2, intColumnV2, stringColumnV2, unsortableColumnV2`. And please rename `ElementColumn` into `CustomColumn`.

Regarding the wrapped El. Yes, you will have to wrap it in an Element.

lukasturcani commented:
Ok updated:

# Description

* Added new module `Internal.SortTableV2` which has modified 
* `sortTableV2` also uses the style, `SortTableStyle` from `Internal.SortTable`
* Functions in `Internal.SortTableV2` are copies of functions from `Internal.SortTable` but with type signatures modified to be polymorphic on the `msg` type. The `sortTableV2` function is slightly modified to work with the `CustomColumn`.
* Added `Page.SortTableV2` and updated `explorer/src/Main.elm` to use it
* Added an png of the new sort table in `docs/assets/sortTableV2.png`
* Custom elements provided by the user are automatically placed into an `Element.el` to prevent scaling issues by default

lukasturcani commented:
Also updated `docs.json`
